flatsurf/geometry/similarity_surface_generators.py

```diff
@@ -1145,6 +1145,227 @@ def mcmullen_genus2_prototype(w, h, t, e, rel=0, base_ring=None):
+        s = Surface_list(base_ring=K)
+        s.add_polygon(
+            polygons(
+                vertices=[(0, 0), (λ, 0), (w, 0), (w + t, h), (λ + t, h), (t, h)],
+                ring=K,
+            )
+        )
+        s.add_polygon(polygons(vertices=[(0, 0), (λ, 0), (λ, λ), (0, λ)], ring=K))
+        s.add_polygon(
+            polygons(
+                vertices=[
+                    (0, 0),
+                    (w - λ, 0),
+                    (w, 0),
+                    (w + t, h),
+                    (w - λ + t, h),
+                    (t, h),
+                ],
+                ring=K,
+            )
+        )
+        s.set_edge_pairing(0, 0, 2, 3)
+        s.set_edge_pairing(0, 1, 0, 3)
+        s.set_edge_pairing(0, 2, 0, 5)
+        s.set_edge_pairing(0, 4, 1, 0)
+        s.set_edge_pairing(1, 1, 1, 3)
+        s.set_edge_pairing(1, 2, 2, 1)
+        s.set_edge_pairing(2, 0, 2, 4)
+        s.set_edge_pairing(2, 2, 2, 5)
+        s.set_immutable()
+        return TranslationSurface(s)
+
+    @staticmethod
+    def lanneau_nguyen_genus4_prototype(w, h, t, e):
+        r"""
+        Return the Lanneau--Ngyuen prototype in the stratum H(6) with
+        parameters ``w``, ``h``, ``t``, and ``e``.
+
+        The associated discriminant is `D = e^2 + 4 wh`.
+
+        INPUT:
+
+        - ``w`` -- a positive integer
+
+        - ``h`` -- a positive integer
+
+        - ``t`` -- a non-negative integer strictly smaller than the gcd of
+          ``w`` and ``h``
+
+        - ``e`` -- a positive integer strictly smaller than `2w - \sqrt{D}`
+
+        EXAMPLES::
+
+            sage: from flatsurf import translation_surfaces
+            sage: T = translation_surfaces.lanneau_nguyen_genus4_prototype(1,1,0,-2)
+            sage: T.stratum()
+            H_4(6)
+
+        REFERENCES:
+
+        - Erwan Lanneau, and Duc-Manh Nguyen, "Weierstrass Prym eigenforms in
+          genus four", Journal of the Institute of Mathematics of Jussieu,
+          19.6, 2020, pp.2045-2085
+
+        """
+        from sage.all import gcd
+
+        w = ZZ(w)
+        h = ZZ(h)
+        t = ZZ(t)
+        e = ZZ(e)
+
+        if not w > 0:
+            raise ValueError("w must be positive")
+        if not h > 0:
+            raise ValueError("h must be positive")
+        if not t >= 0:
+            raise ValueError("t must be non-negative")
+        if not t < gcd(w, h):
+            raise ValueError("t must be smaller than the gcd of w and h")
+        if not gcd([w, h, t, e]) == 1:
+            raise ValueError("w, h, t, e must be coprime")
+
+        K = QQ
+        D = K(e**2 + 4 * w * h)
+
+        if not D.is_square():
+            from sage.all import QuadraticField
+
+            K = QuadraticField(D)
+            D = K(D)
+
+        d = D.sqrt()
+
+        λ = (e + d) / 2
+
+        if not λ < w:
+            raise ValueError("λ must be smaller than w")
+
+        if λ == w / 2:
+            raise ValueError("λ and w/2 must be distinct")
```

Comment:
Not sure if you had these conditions in your implementation? I took them from the article.